src/DataStructures/DynamicBuffer.hpp

```diff
@@ -0,0 +1,105 @@
+
+#include "DataStructures/DataVector.hpp"
+#include "Utilities/TypeTraits/GetFundamentalType.hpp"
+#include "Utilities/TypeTraits/IsIterable.hpp"
+/**
+ * \ingroup DataStructuresGroup
+ *
+ * \brief A dynamically sized vector of DataVectors or ComplexDataVectors. If T
+ * is a floating point type, `DynamicBuffer<T>` resolves to std::vector<T>.
+ *
+ * \details This class is useful when one wants to create a `std::vector<T>`
+ * with a size that is unknown at compile time. It allocates all DataVectors in
+ * a single memory chunk rather than allocating each individually. If the size
+ * of the vector is known at compile time, a `TempBuffer` object should be used
+ * instead.
+ */
+
+template <typename T>
+class DynamicBuffer {
+ public:
+  static constexpr bool is_data_vector_type =
+      std::is_base_of_v<VectorImpl<tt::get_fundamental_type_t<T>, T>, T>;
+
+  DynamicBuffer() = default;
+
+  /*
+   * Constructs a `DynamicBuffer`. The `outer_size` is the size of the outer
+   * `std::vector`, the `inner_size` is the size of the objects of type T stored
+   * inside the DataVector. The `inner_size` has to be size 1 if T is a
+   * fundamental type.
+   */
+  DynamicBuffer(const size_t outer_size, const size_t inner_size) noexcept
+      : inner_size_(inner_size), data_vectors_(outer_size) {
+    if constexpr (is_data_vector_type) {
+      buffer_.resize(outer_size * inner_size_);
+      set_references();
+    } else {
+      if (inner_size != 1) {
+        ERROR(
+            "DynamicBuffer must have inner_size == 1 when T is a fundamental "
+            "type but has inner size "
+            << inner_size_);
+      }
+    }
+  }
+
+  ~DynamicBuffer() = default;
+  DynamicBuffer(const DynamicBuffer& other) noexcept
+      : inner_size_(other.inner_size_) {
+    if constexpr (is_data_vector_type) {
+      data_vectors_.resize(other.size());
+      buffer_ = other.buffer_;
+      set_references();
+    } else {
+      data_vectors_ = other.data_vectors_;
+    }
+  }
+  DynamicBuffer(DynamicBuffer&& other) noexcept
+      : inner_size_(other.inner_size_) {
+    if constexpr (is_data_vector_type) {
+      data_vectors_.resize(other.size());
+      buffer_ = std::move(other.buffer_);
+      set_references();
+    } else {
+      data_vectors_ = std::move(other.data_vectors_);
+    }
+  }
+
+  DynamicBuffer& operator=(DynamicBuffer other) noexcept {
+    std::swap(*this, other);
+    return *this;
+  }
+
+  T& operator[](size_t index) { return data_vectors_[index]; }
+  T& at(size_t index) { return data_vectors_.at(index); }
+  const T& operator[](size_t index) const { return data_vectors_[index]; }
+  const T& at(size_t index) const { return data_vectors_.at(index); }
+
+  auto begin() { return data_vectors_.begin(); }
+  auto end() { return data_vectors_.end(); }
+  auto begin() const { return data_vectors_.begin(); }
+  auto end() const { return data_vectors_.end(); }
+
+  size_t size() const noexcept { return data_vectors_.size(); }
+
```

Comment:
Can you make this buffer serializable? (add a `pup` + test) We basically need everything to be serializable and it's annoying to discover something is missing a feature like `pup` and then having to add it later